### DISCUSSION TITLE
pcpcombine and UFS UPP output

### DISCUSSION THREAD

**Discussion by benjamin-cash** [dtcenter/METplus#1975] [Configuration] [answered]
Hello,

I am trying to use pcpcombine to generate daily/weekly data from inline post output from UFS. The run produces a series of files labeled:

GFSPRS.GrbF00
GFSPRS.GrbF06
GFSPRS.GrbF12
...
GFSPRS.GrbF102
GFSPRS.GrbF108

There are also a number of other files in the directory that require additional preprocessing that I don't want to analyze yet. Is there a sample configuration file specifically for this kind of analysis for UFS/UPP that someone could point me to? Thanks!

**Comment by georgemccabe** [comment, marked as answer]
Hi @benjamin-cash, I don't know that we have any examples that process UFS/UPP data specifically. The [Grid-Stat: 24-hour QPF Use Case](https://metplus.readthedocs.io/en/latest/generated/model_applications/precipitation/GridStat_fcstGFS_obsCCPA_GRIB.html) is similar in that it uses PCPCombine to build daily accumulations from 6 hourly GRIB input files. The FCST_PCP_COMBINE_INPUT_TEMPLATE for your data would look something like this:

`FCST_PCP_COMBINE_INPUT_TEMPLATE = GFSPRS.GrbF{lead?fmt=%H}`

**Reply by benjamin-cash** [reply]
Hi @georgemccabe - thanks! I've tried following that example, but when I do I run into an error. I have:

LOOP_BY = VALID
VALID_TIME_FMT = %Y%m%d%H
VALID_BEG = 2012010100
VALID_END = 2012010100
VALID_INCREMENT = 86400
LEAD_SEQ = 24

but when I do this I get:
12/02 22:16:12.106Z metplus INFO: COMMAND: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/pcp_combine -v 2 -sum 20111231_000000 06 20120101_000000 24 -pcpdir /scratch1/02441/bcash/prototype-p8/two_year_ensemble.744.x12.y16.cpn36.wg3.wt108/run/2012010100/C384/mem001 -pcprx GFSPRS\.GrbF.* -name "A001500" /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/NEWSe24H_mem000_lag00.nc
OUTPUT:
DEBUG 1: Start pcp_combine by bcash(817373) at 2022-12-02 22:16:12Z  cmd: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/pcp_combine -v 2 -sum 20111231_000000 06 20120101_000000 24 -pcpdir /scratch1/02441/bcash/prototype-p8/two_year_ensemble.744.x12.y16.cpn36.wg3.wt108/run/2012010100/C384/mem001 -pcprx GFSPRS.GrbF.* -name A001500 /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/NEWSe24H_mem000_lag00.nc
DEBUG 2: Performing sum command: Init/In_Accum/Valid/Out_Accum Times = 20111231_000000/060000/20120101_000000/240000
DEBUG 2: Searching for 4 files with accumulation times of 060000 to sum to a total accumulation time of 240000.
ERROR  :
ERROR  : sum_data_files() -> cannot find a file with a valid time of 20120101_000000 and accumulation time of 060000 matching the regular expression "GFSPRS.GrbF.*"

My model is initialized 2012010100, so it isn't surprising that it is throwing an error if it is looking for a file with a time of 20111231_000000. In the example though VALID_BEG is set to the init time, so I am not sure what I am doing wrong here.

**Reply by georgemccabe** [reply]
I didn't realize that this use case uses `FCST_PCP_COMBINE_METHOD = SUM`. I would recommend changing this to `FCST_PCP_COMBINE_METHOD = ADD` and see if that helps.

**Reply by georgemccabe** [reply]
You may also need to set `FCST_PCP_COMBINE_CONSTANT_INIT = True`

**Reply by benjamin-cash** [reply]
Thanks, that at least got me past the starting before the valid time problem. :) It looks like it processed the first 24 hours of the run, what do I need to change to get it to process more days? Do I increase VALID_END to my targeted end date?

Hmm... apparently not. It performed two calculations for two different valid times but selected the same files to add each time.

Hmm x2... changing VALID_BEG and VALID_END doesn't affect the files chosen to add at all, so I am clearly missing some piece of understanding here. I will pause and wait for further guidance. :)

**Reply by georgemccabe** [reply]
The files only have forecast lead information in their names, so you will have to include the year/month/day of the time to find the correct files. It looks like there are directories that are in YYYYMMDDHH format, so you will have to include that in your template.

```
FCST_PCP_COMBINE_INPUT_DIR = /scratch1/02441/bcash/prototype-p8/two_year_ensemble.744.x12.y16.cpn36.wg3.wt108/run
FCST_PCP_COMBINE_INPUT_TEMPLATE = {init?fmt=%Y%m%d%H}/C384/mem001/GFSPRS.GrbF{lead?fmt=%H}
```

**Reply by benjamin-cash** [reply]
Could not find files to build accumulation in /scratch1/02441/bcash/prototype-p8/two_year_ensemble.744.x12.y16.cpn36.wg3.wt108/run using template {init?fmt=%Y%m%d%H}/C384/mem001/GFSPRS.GrbF{lead?fmt=%H}

Do I need to specify init somewhere?

**Reply by georgemccabe** [reply]
It will calculate init for you using the valid time and forecast lead. For a valid time of 2012010100 and 24 hour lead, it will be looking for an init directory for 2011123100. Is the data for Dec. 31, 2011 available? If not, you may have to adjust your VALID_BEG time to start on the 2nd instead.

**Reply by benjamin-cash** [reply]
Hi @georgemccabe - The run is initialized on 2012010100, so there is no data before that. So it is a little counterintuitive to put VALID_BEG as 2012010200 but I can try that. 

More generally, this is a seasonal run that is initialized on 2012010100 and will run for several months. What I am hoping to do with pcpcombine is to take the 6h grib output and generate a time series of daily accumulations. Is this possible?

**Reply by georgemccabe** [reply]
Hi @benjamin-cash, you can configure the use case to loop by initialization time instead of valid time.
You can still reference the valid time in the output files because it will be calculated. If the init time will be constant for each run, you should be able to set a single time (INIT_BEG == INIT_END) and define a list of forecast leads to use for each day:

```
LOOP_BY = INIT
INIT_TIME_FMT = %Y%m%d%H
INIT_BEG = 2012010100
INIT_END = 2012010100
INIT_INCREMENT = 1d
LEAD_SEQ = 1d, 2d, 3d, 4d
```

You should be able to use [SeriesAnalysis ](https://metplus.readthedocs.io/en/latest/Users_Guide/wrappers.html#seriesanalysis)to generate a time series of the daily accumulations.

**Reply by benjamin-cash** [reply]
So in LEAD_SEQ, if I have a 180 day run would I list out to 180d?

**Reply by georgemccabe** [reply]
Yes. You can use the begin_end_incr syntax to easily define the list of values.

`LEAD_SEQ = begin_end_incr(1,180,1)d`

**Reply by benjamin-cash** [reply]
@georgemccabe - Fantastic. It looks like that worked to process multiple days, but it seems like it wrote everything to the same output file. Is there a way to control the name of the output file consistent with the LEAD_SEQ value?

**Reply by georgemccabe** [reply]
Yes, if you reference the valid time in the output template, it will produce a different filename for each run:

`FCST_PCP_COMBINE_OUTPUT_TEMPLATE = {valid?fmt=%Y%m%d%H}.nc`

**Reply by benjamin-cash** [reply]
That seems to have done the trick, many thanks for all of the help! Is there a use case or example for using SeriesAnalysis to create the timeseries? Just looking at the page you pointed me to was a bit overwhelming. :)

**Reply by georgemccabe** [reply]
The [SeriesAnalysis; Basic Use Case](https://metplus.readthedocs.io/en/latest/generated/met_tool_wrapper/SeriesAnalysis/SeriesAnalysis.html#sphx-glr-generated-met-tool-wrapper-seriesanalysis-seriesanalysis-py) is an example that may help. Let me know if you have any questions about that example.

**Reply by benjamin-cash** [reply]
Hi @georgemccabe - I'm getting an error message that just reads "ERROR  : Trouble reading input data files.". I'm not really sure where to go from there.

**Reply by georgemccabe** [reply]
Hi @benjamin-cash, could you provide the full log file?

**Reply by benjamin-cash** [reply]
12/05 22:06:58.111Z metplus INFO: ****************************************
12/05 22:06:58.112Z metplus INFO: * Running METplus SeriesAnalysisWrapper
12/05 22:06:58.112Z metplus INFO: *  at init time: 2012-01-01 00:00
12/05 22:06:58.112Z metplus INFO: ****************************************
12/05 22:06:58.112Z metplus INFO: Processing forecast lead 1 day
12/05 22:06:58.116Z metplus INFO: Setting env variables from config file...
12/05 22:06:58.116Z metplus INFO: ENVIRONMENT FOR NEXT COMMAND:
12/05 22:06:58.116Z metplus INFO: CLIMO_MEAN_FILE=
12/05 22:06:58.116Z metplus INFO: CLIMO_STDEV_FILE=
12/05 22:06:58.116Z metplus INFO: FCST_FIELD={ name="A001500"; level="24h"; }
12/05 22:06:58.116Z metplus INFO: FCST_FILE_TYPE=
12/05 22:06:58.116Z metplus INFO: METPLUS_BLOCK_SIZE=
12/05 22:06:58.116Z metplus INFO: METPLUS_CAT_THRESH=
12/05 22:06:58.116Z metplus INFO: METPLUS_CLIMO_CDF_DICT=
12/05 22:06:58.116Z metplus INFO: METPLUS_CLIMO_MEAN_DICT=
12/05 22:06:58.117Z metplus INFO: METPLUS_CLIMO_STDEV_DICT=
12/05 22:06:58.117Z metplus INFO: METPLUS_CTS_LIST=
12/05 22:06:58.117Z metplus INFO: METPLUS_DESC=
12/05 22:06:58.117Z metplus INFO: METPLUS_FCST_CAT_THRESH=
12/05 22:06:58.117Z metplus INFO: METPLUS_FCST_FIELD=field = [{ name="A001500"; level="24h"; }];
12/05 22:06:58.117Z metplus INFO: METPLUS_FCST_FILE_TYPE=
12/05 22:06:58.117Z metplus INFO: METPLUS_HSS_EC_VALUE=
12/05 22:06:58.117Z metplus INFO: METPLUS_MET_CONFIG_OVERRIDES=
12/05 22:06:58.117Z metplus INFO: METPLUS_MODEL=model = "GFS";
12/05 22:06:58.117Z metplus INFO: METPLUS_OBS_CAT_THRESH=
12/05 22:06:58.117Z metplus INFO: METPLUS_OBS_FIELD=field = [{ name="A001500"; level="24h"; }];
12/05 22:06:58.117Z metplus INFO: METPLUS_OBS_FILE_TYPE=
12/05 22:06:58.117Z metplus INFO: METPLUS_OBTYPE=obtype = "ANLYS";
12/05 22:06:58.117Z metplus INFO: METPLUS_OUTPUT_STATS_DICT=
12/05 22:06:58.117Z metplus INFO: METPLUS_REGRID_DICT=
12/05 22:06:58.117Z metplus INFO: METPLUS_STAT_LIST=
12/05 22:06:58.117Z metplus INFO: METPLUS_VLD_THRESH=
12/05 22:06:58.117Z metplus INFO: MET_TMP_DIR=/work2/02441/bcash/frontera/METplus/output/tmp
12/05 22:06:58.117Z metplus INFO: MODEL=GFS
12/05 22:06:58.117Z metplus INFO: OBS_FIELD={ name="A001500"; level="24h"; }
12/05 22:06:58.117Z metplus INFO: OBS_FILE_TYPE=
12/05 22:06:58.117Z metplus INFO: OBTYPE=ANLYS
12/05 22:06:58.117Z metplus INFO: OMP_NUM_THREADS=1
12/05 22:06:58.117Z metplus INFO: REGRID_TO_GRID=NONE
12/05 22:06:58.117Z metplus INFO: STAT_LIST=[]
12/05 22:06:58.119Z metplus INFO: COMMAND: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
OUTPUT:
DEBUG 1: Start series_analysis by bcash(817373) at 2022-12-05 22:06:58Z  cmd: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
DEBUG 1: Default Config File: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/share/met/config/SeriesAnalysisConfig_default
DEBUG 1: User Config File: /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped
ERROR  :
ERROR  : Trouble reading input data files.
ERROR  :
12/05 22:06:58.491Z metplus (command_builder.py:263) ERROR: (command_builder.py:1275) MET command returned a non-zero return code:/work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
12/05 22:06:58.492Z metplus INFO: Check the logfile for more information on why it failed: /work2/02441/bcash/frontera/METplus/output/logs/metplus.log.20221205220650
12/05 22:06:58.492Z metplus INFO: Processing forecast lead 2 days
12/05 22:06:58.494Z metplus INFO: Setting env variables from config file...
12/05 22:06:58.494Z metplus INFO: ENVIRONMENT FOR NEXT COMMAND:
12/05 22:06:58.494Z metplus INFO: CLIMO_MEAN_FILE=
12/05 22:06:58.494Z metplus INFO: CLIMO_STDEV_FILE=
12/05 22:06:58.494Z metplus INFO: FCST_FIELD={ name="A001500"; level="24h"; }
12/05 22:06:58.494Z metplus INFO: FCST_FILE_TYPE=
12/05 22:06:58.494Z metplus INFO: METPLUS_BLOCK_SIZE=
12/05 22:06:58.494Z metplus INFO: METPLUS_CAT_THRESH=
12/05 22:06:58.495Z metplus INFO: METPLUS_CLIMO_CDF_DICT=
12/05 22:06:58.495Z metplus INFO: METPLUS_CLIMO_MEAN_DICT=
12/05 22:06:58.495Z metplus INFO: METPLUS_CLIMO_STDEV_DICT=
12/05 22:06:58.495Z metplus INFO: METPLUS_CTS_LIST=
12/05 22:06:58.495Z metplus INFO: METPLUS_DESC=
12/05 22:06:58.495Z metplus INFO: METPLUS_FCST_CAT_THRESH=
12/05 22:06:58.495Z metplus INFO: METPLUS_FCST_FIELD=field = [{ name="A001500"; level="24h"; }];
12/05 22:06:58.495Z metplus INFO: METPLUS_FCST_FILE_TYPE=
12/05 22:06:58.495Z metplus INFO: METPLUS_HSS_EC_VALUE=
12/05 22:06:58.495Z metplus INFO: METPLUS_MET_CONFIG_OVERRIDES=
12/05 22:06:58.495Z metplus INFO: METPLUS_MODEL=model = "GFS";
12/05 22:06:58.495Z metplus INFO: METPLUS_OBS_CAT_THRESH=
12/05 22:06:58.495Z metplus INFO: METPLUS_OBS_FIELD=field = [{ name="A001500"; level="24h"; }];
12/05 22:06:58.495Z metplus INFO: METPLUS_OBS_FILE_TYPE=
12/05 22:06:58.495Z metplus INFO: METPLUS_OBTYPE=obtype = "ANLYS";
12/05 22:06:58.495Z metplus INFO: METPLUS_OUTPUT_STATS_DICT=
12/05 22:06:58.495Z metplus INFO: METPLUS_REGRID_DICT=
12/05 22:06:58.495Z metplus INFO: METPLUS_STAT_LIST=
12/05 22:06:58.495Z metplus INFO: METPLUS_VLD_THRESH=
12/05 22:06:58.495Z metplus INFO: MET_TMP_DIR=/work2/02441/bcash/frontera/METplus/output/tmp
12/05 22:06:58.495Z metplus INFO: MODEL=GFS
12/05 22:06:58.495Z metplus INFO: OBS_FIELD={ name="A001500"; level="24h"; }
12/05 22:06:58.495Z metplus INFO: OBS_FILE_TYPE=
12/05 22:06:58.495Z metplus INFO: OBTYPE=ANLYS
12/05 22:06:58.495Z metplus INFO: OMP_NUM_THREADS=1
12/05 22:06:58.495Z metplus INFO: REGRID_TO_GRID=NONE
12/05 22:06:58.495Z metplus INFO: STAT_LIST=[]
12/05 22:06:58.497Z metplus INFO: COMMAND: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
OUTPUT:
DEBUG 1: Start series_analysis by bcash(817373) at 2022-12-05 22:06:58Z  cmd: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
DEBUG 1: Default Config File: /work2/06612/tg859120/frontera/MET/11.0.0-beta3/share/met/config/SeriesAnalysisConfig_default
DEBUG 1: User Config File: /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped
ERROR  :
ERROR  : Trouble reading input data files.
ERROR  :
12/05 22:06:58.701Z metplus (command_builder.py:263) ERROR: (command_builder.py:1275) MET command returned a non-zero return code:/work2/06612/tg859120/frontera/MET/11.0.0-beta3/bin/series_analysis -both /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/BOTH_FILES -out /work2/02441/bcash/frontera/METplus/output/met_tool_wrapper/PCPCombine/PCPCombine_sum/ -config /work2/06612/tg859120/frontera/METplus/METplus-5.0.0-beta3/parm/met_config/SeriesAnalysisConfig_wrapped -v 2
12/05 22:06:58.702Z metplus INFO: Check the logfile for more information on why it failed: /work2/02441/bcash/frontera/METplus/output/logs/metplus.log.20221205220650
12/05 22:06:58.704Z metplus.SeriesAnalysis (met_util.py:170) ERROR: SeriesAnalysis had 2 errors.
12/05 22:06:58.706Z metplus INFO: Overwrite final conf here: /work2/02441/bcash/frontera/METplus/output/metplus_final.conf.20221205220650
12/05 22:06:58.708Z metplus (met_util.py:218) ERROR: METplus has finished running as user bcash(817373) but had 2 errors.
12/05 22:06:58.708Z metplus INFO: Check the log file for more information: /work2/02441/bcash/frontera/METplus/output/logs/metplus.log.20221205220650

**Reply by benjamin-cash** [reply]
@georgemccabe is there a way to have netcdf files written out with a record dimension by pcpcombine? If that's the case then I could just use ncrcat to stitch the files together.

**Reply by georgemccabe** [reply]
Could you set `LOG_LEVEL = DEBUG` and rerun so I can see which files it is gathering? Also, the conf file you used would be helpful.

From a glance here, it looks like you have it configured to run using BOTH instead of defining separate FCST and OBS files. This is typically used when the forecast and observation data are found in the same file, such as GridStat matched pairs NetCDF output. The output from PCPCombine will only contain forecast output, so you will need a corresponding observation data files to process.

**Reply by benjamin-cash** [reply]
Hi @georgemccabe - I was wondering about the BOTH setting, but it gave me an error without it. In this case I don't have any observational data I want to compare to, I was just trying to generate a time series. Can I use this to just process model files without observations?

**Reply by georgemccabe** [reply]
Hi @benjamin-cash, there isn't a MET tool to accomplish this, but as you mentioned there are some useful NCO tools that can be used to combine the data into a single file. The MET tools always write out data with 2 dimensions, but you should still be able to combine them. I'm not sure if the exact command or tool offhand, but I could do some digging after some meetings. Please let me know if you figure out how to do this and what commands you used to do it. That info would likely be useful to other users.

**Reply by benjamin-cash** [reply]
Hi @georgemccabe - Sure, I'll post the recipe here when I work out the details. Just to be clear, was this answer also to the question of whether SeriesAnalysis can combine just the model files?

**Reply by georgemccabe** [reply]
Yes. SeriesAnalysis compares time series of 2 different gridded datasets, generates matched pair differences (i.e. fcst - obs), and computes statistics to summarize the comparison of the two datasets. More information can be found in the [Series-Analysis chapter](https://met.readthedocs.io/en/latest/Users_Guide/series-analysis.html) of the MET User's Guide.

**Reply by benjamin-cash** [reply]
Gotcha. Not really a good fit for my use case then - I'll go ahead and close this out. Thanks again for all the help!
